src/core/execute.c

```diff
@@ -1934,16 +1944,90 @@ static int setup_exec_directory(
                 if (r < 0)
                         goto fail;
 
-                r = mkdir_p_label(p, context->directories[type].mode);
-                if (r < 0)
+                if (context->dynamic_user && type != EXEC_DIRECTORY_CONFIGURATION) {
```

Comment:
Is this logic also applied to `RuntimeDirectory=`? I think `EXEC_DIRECTORY_RUNTIME` should also be excluded.

Reply:
Ahh... There is `RuntimeDirectoryPreserve=` option. What do you think about that we apply the logic for `RuntimeDirectory=` if `RuntimeDirectoryPreserve=` is enabled?

Reply:
hmm? but we do apply the logic for EXEC_DIRECTORY_RUNTIME!

not grokking what you are saying?

Reply:
Sorry for my misleading comment. I just want to say that we may not need to apply the logic if `RuntimeDirectoryPreserve=no`. Because in such a case the runtime directories are always removed, then we do not need to care whether the uid owning the directories conflicts or not.

Reply:
so, i thought about that, but I was unsure whether that's worth it. I mean, there's the risk that the removal of the runtime directory does not work, due to a broken selinux policy (or some other MAC issue or so we don't expect). Now, if the removal didn't work then we'd have a directory directly sitting in /run that is still owned by one of the dynamic UIDs. If that happens the UID becomes "toxic": we'd need to make sure we never use it again until the subdir is properly removed eventually.  But making sure of that would need some non-trivial amount of logic: either we'd have to take note of this somewhere, and maintain a list of toxic UIDs not to hand out again, or we'd have to enumerate through /run whenever we assign a new dynamic UID to see if there's something owned by that UID, but that'd be kinda slow and ugly.

Hence I figured let's avoid this: use the `private/` subdirectory trick also in this case. That way having the dir lying around doesn't matter if its removal failed, as the directory won't be accessible from other services (or the host) anyway...

does that make sense?

Reply:
Thank you for the detailed explanation. Now I agree with you.

Reply:
This decision has caused problems. See https://github.com/systemd/systemd/issues/7761